components/connectionmanager/src/eviction.rs

```diff
@@ -0,0 +1,818 @@
+                    assert_eq!(filtered_counter.len(), 2);
+                }
+                8 => {
+                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
+                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
+                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
+                    assert_eq!(removed_counter.len(), 3);
+
+                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
+                    assert_eq!(filtered_counter.len(), 1);
+                }
+                9 => {
+                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
+                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
+                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
+                    assert_eq!(removed_counter[&7.0_f64.to_bits()], 1);
+                    assert_eq!(removed_counter.len(), 4);
+
+                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 1);
+                    assert_eq!(filtered_counter.len(), 1);
+                }
+                10 => {
+                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
+                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
+                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
+                    assert_eq!(removed_counter[&7.0_f64.to_bits()], 2);
+                    assert_eq!(removed_counter.len(), 4);
+
+                    assert_eq!(filtered_counter.len(), 0);
+                }
+                _ => panic!("unexpected i value"),
+            }
+        }
+    }
+
+    #[test]
+    fn test_eviction_iter_select_peers_weighted() {
+        try_init_logger("info");
+        let test_peers = build_test_peers();
+        let test_peers = test_peers.iter().collect::<Vec<_>>();
+        let eviction_iter_vec = EvictionIter::from_peers(&test_peers).collect::<Vec<(&Peer, EvictionRanks)>>();
+        let total_weight = 59.5;
+        let expected_probabilities = vec![
+            // we add one to avoid 0, and nan numbers.. `weight_strats::by_highest_none_latency_rank` adds 1 to the rank for these hypothetical situation.
+            (1.0 + 1.0) / total_weight,
+            (1.5 + 1.0) / total_weight,
+            (2.0 + 1.0) / total_weight,
+            (4.0 + 1.0) / total_weight * 2.0, // we have two 4.0 ranks
+            (5.0 + 1.0) / total_weight,
+            (7.5 + 1.0) / total_weight * 2.0, // we have two 7.5 ranks
+            (8.0 + 1.0) / total_weight,
+            (9.0 + 1.0) / total_weight,
+        ];
+        assert_eq!(expected_probabilities.iter().sum::<f64>(), 1.0);
+        assert_eq!(expected_probabilities.len(), 8);
+
+        let mut selected_counter = HashMap::<u64, f64>::new();
+        let num_of_trials = 2054;
+        for _ in 0..num_of_trials {
+            //println!("sample_size: {}", sample_size);
+            let eviction_iter = Box::new(eviction_iter_vec.clone().into_iter()) as EvictionIter;
```

Comment:
```suggestion
pub trait EvictionIterExt<'a, T: Iterator<Item = (&'a Peer, EvictionRanks)> + 'a>: IntoIterator<Item = (&'a Peer, EvictionRanks), IntoIter = T>{
    fn filter_peers<F>(self, amount: usize, compare_fn: F) -> impl Iterator<Item = (&'a Peer, EvictionRanks)> + 'a
    where
        F: Fn(&EvictionRanks, &EvictionRanks) -> Ordering,
        Self: Sized,
    {
        let rng = &mut thread_rng();
        self.into_iter().sorted_unstable_by(move |(_, r1), (_, r2)| match compare_fn(r1, r2) {
            Ordering::Greater => Ordering::Greater,
            Ordering::Less => Ordering::Less,
            // we tie break randomly, as to not expose preference due to pre-existing ordering.
            Ordering::Equal => {
                if rng.gen_bool(0.5) {
                    Ordering::Greater
                } else {
                    Ordering::Less
                }
            }
        })
        .skip(amount)
    }

    fn select_peers_weighted<F>(self, amount: usize, weight_fn: F) -> impl Iterator<Item = (&'a Peer, EvictionRanks)> + 'a
    where
        F: Fn(&EvictionRanks) -> f64,
        Self: Sized,
    {
        let rng = &mut thread_rng();
        self.into_iter().collect_vec().choose_multiple_weighted(rng, amount, |(_, r)| weight_fn(r)).unwrap().copied().collect_vec().into_iter()
    }

    fn iterate_peers(self) -> impl Iterator<Item = &'a Peer> + 'a
    where
        Self: Sized,
    {
        self.into_iter().map(|(p, _)| p)
    }
}

impl<'a, T: Iterator<Item = (&'a Peer, EvictionRanks)> + 'a, IntoIter: IntoIterator<Item = (&'a Peer, EvictionRanks), IntoIter = T>> EvictionIterExt<'a, T> for IntoIter {}

pub fn from_peers<'a>(peers: &'a [&'a Peer]) -> impl Iterator<Item = (&'a Peer, EvictionRanks)> + 'a {
    let ip_prefix_map = build_ip_prefix_map(peers);
    let mut ranks = vec![EvictionRanks::default(); peers.len()];
    Box::new(peers.iter().enumerate().map(move |(i1, p1)| {
        for (i2, p2) in peers[i1..].iter().enumerate().skip(1) {
            match ip_prefix_map[&p1.prefix_bucket()].cmp(&ip_prefix_map[&p2.prefix_bucket()]) {
                // low is good, so we add rank to the peer with the greater ordering.
                Ordering::Greater => ranks[i1].ip_prefix_bucket += 1.0,
                Ordering::Less => ranks[i1 + i2].ip_prefix_bucket += 1.0,
                Ordering::Equal => {
                    ranks[i1].ip_prefix_bucket += 0.5;
                    ranks[i1 + i2].ip_prefix_bucket += 0.5;
                }
            };

            match p1.time_connected().cmp(&p2.time_connected()) {
                // high is good, so we add to the peer with the lesser ordering.
                Ordering::Greater => ranks[i1 + i2].time_connected += 1.0,
                Ordering::Less => ranks[i1].time_connected += 1.0,
                Ordering::Equal => {
                    ranks[i1].time_connected += 0.5;
                    ranks[i1 + i2].time_connected += 0.5;
                }
            };

            match p1.last_ping_duration().cmp(&p2.last_ping_duration()) {
                // low is good, so we add to the peer with the greater ordering.
                Ordering::Greater => ranks[i1].last_ping_duration += 1.0,
                Ordering::Less => ranks[i1 + i2].last_ping_duration += 1.0,
                Ordering::Equal => {
                    ranks[i1].last_ping_duration += 0.5;
                    ranks[i1 + i2].last_ping_duration += 0.5;
                }
            };

            match (p1.last_block_transfer(), p2.last_block_transfer()) {
                // Some is good, so we add to the peer with None
                (Some(_), None) => ranks[i1 + i2].last_block_transfer += 1.0,
                (None, Some(_)) => ranks[i1].last_block_transfer += 1.0,
                (None, None) => {
                    ranks[i1].last_block_transfer += 0.5;
                    ranks[i1 + i2].last_block_transfer += 0.5;
                }
                (Some(peer1_last_block_transfer), Some(peer2_last_block_transfer)) => {
                    match peer1_last_block_transfer.cmp(&peer2_last_block_transfer) {
                        // low is good, so we add to the peer with the greater ordering.
                        Ordering::Greater => ranks[i1].last_block_transfer += 1.0,
                        Ordering::Less => ranks[i1 + i2].last_block_transfer += 1.0,
                        Ordering::Equal => {
                            ranks[i1].last_block_transfer += 0.5;
                            ranks[i1 + i2].last_block_transfer += 0.5;
                        }
                    }
                }
            };

            match (p1.last_tx_transfer(), p2.last_tx_transfer()) {
                // Some is good, so we add to the peer with None
                (Some(_), None) => ranks[i1 + i2].last_tx_transfer += 1.0,
                (None, Some(_)) => ranks[i1].last_tx_transfer += 1.0,
                (None, None) => {
                    ranks[i1].last_tx_transfer += 0.5;
                    ranks[i1 + i2].last_tx_transfer += 0.5;
                }
                (Some(peer1_last_tx_transfer), Some(peer2_last_tx_transfer)) => {
                    match peer1_last_tx_transfer.cmp(&peer2_last_tx_transfer) {
                        // low is good, so we add to the peer with the greater ordering.
                        Ordering::Greater => ranks[i1].last_tx_transfer += 1.0,
                        Ordering::Less => ranks[i1 + i2].last_tx_transfer += 1.0,
                        Ordering::Equal => {
                            ranks[i1].last_tx_transfer += 0.5;
                            ranks[i1 + i2].last_tx_transfer += 0.5;
                        }
                    }
                }
            };
        }
        (peers[i1], ranks[i1])
    }))
}
// Abstracted helper functions:

fn build_ip_prefix_map(peers: &[&Peer]) -> HashMap<PrefixBucket, usize> {
    let mut ip_prefix_map = HashMap::new();
    for peer in peers.iter() {
        *ip_prefix_map.entry(peer.prefix_bucket()).or_insert(1) += 1;
    }
    ip_prefix_map
}

#[cfg(test)]
mod test {

    use super::*;

    use kaspa_core::{debug, info, log::try_init_logger};
    use kaspa_p2p_lib::Peer;
    use kaspa_utils::networking::PeerId;
    use std::net::SocketAddr;
    use std::{
        net::SocketAddrV4,
        str::FromStr,
        time::{Duration, Instant},
    };
    use uuid::Uuid;

    fn build_test_peers() -> Vec<Peer> {
        let now = Instant::now();
        let instants = Vec::from_iter((0..10).map(|i| now.checked_sub(Duration::from_secs(i)).unwrap())); // `from_secs` is important here, as time_connected has ms granularity, so it most be greater granularity than ms.
                                                                                                          // rank 0, 1, 2, 3, 5, => 0, 1
        let peer0 = Peer::new(
            PeerId::from(Uuid::from_u128(0u128)),
            SocketAddr::V4(SocketAddrV4::from_str("1.0.0.0:1").unwrap()),
            Default::default(),
            instants[8],
            Default::default(),
            2,
            Some(instants[4]),
            Some(instants[3]),
        );
        // rank 1.5, 0, 1, 2, 3 => 0, 1.5
        let peer1 = Peer::new(
            PeerId::from(Uuid::from_u128(1u128)),
            SocketAddr::V4(SocketAddrV4::from_str("2.0.0.0:1").unwrap()),
            Default::default(),
            instants[9],
            Default::default(),
            1,
            Some(instants[5]),
            Some(instants[4]),
        );
        // rank 1.5, 2, 0, 1, 2 => 0, 2
        let peer2 = Peer::new(
            PeerId::from(Uuid::from_u128(2u128)),
            SocketAddr::V4(SocketAddrV4::from_str("2.0.0.0:1").unwrap()),
            Default::default(),
            instants[7],
            Default::default(),
            0,
            Some(instants[6]),
            Some(instants[5]),
        );
        // rank 4, 3, 3, 0, 1 => 0, 4
        let peer3 = Peer::new(
            PeerId::from(Uuid::from_u128(3u128)),
            SocketAddr::V4(SocketAddrV4::from_str("3.0.0.0:1").unwrap()),
            Default::default(),
            instants[6],
            Default::default(),
            3,
            Some(instants[7]),
            Some(instants[6]),
        );
        // rank 4, 4, 4, 4, 0 => 0, 4
        let peer4 = Peer::new(
            PeerId::from(Uuid::from_u128(4u128)),
            SocketAddr::V4(SocketAddrV4::from_str("3.0.0.0:1").unwrap()),
            Default::default(),
            instants[5],
            Default::default(),
            4,
            Some(instants[3]),
            Some(instants[7]),
        );
        // rank 4, 5, 5, 5, 5 => 4, 5
        let peer5 = Peer::new(
            PeerId::from(Uuid::from_u128(5u128)),
            SocketAddr::V4(SocketAddrV4::from_str("3.0.0.0:1").unwrap()),
            Default::default(),
            instants[4],
            Default::default(),
            5,
            Some(instants[2]),
            Some(instants[2]),
        );
        // rank 7.5, 6, 6, 6, 8.5 => 6, 7.5
        let peer6 = Peer::new(
            PeerId::from(Uuid::from_u128(6u128)),
            SocketAddr::V4(SocketAddrV4::from_str("4.0.0.0:1").unwrap()),
            Default::default(),
            instants[3],
            Default::default(),
            6,
            Some(instants[1]),
            None,
        );
        // rank 7.5, 7, 7, 7, 8.5 => 7, 7.5
        let peer7 = Peer::new(
            PeerId::from(Uuid::from_u128(7u128)),
            SocketAddr::V4(SocketAddrV4::from_str("4.0.0.0:1").unwrap()),
            Default::default(),
            instants[2],
            Default::default(),
            7,
            Some(instants[0]),
            None,
        );
        // rank 7.5, 8, 8, 8.5, 6 => 6, 8
        let peer8 = Peer::new(
            PeerId::from(Uuid::from_u128(8u128)),
            SocketAddr::V4(SocketAddrV4::from_str("4.0.0.0:1").unwrap()),
            Default::default(),
            instants[1],
            Default::default(),
            8,
            None,
            Some(instants[1]),
        );
        // rank 7.5, 9, 9, 8.5, 7 => 7, 9
        let peer9 = Peer::new(
            PeerId::from(Uuid::from_u128(9u128)),
            SocketAddr::V4(SocketAddrV4::from_str("4.0.0.0:1").unwrap()),
            Default::default(),
            instants[0],
            Default::default(),
            9,
            None,
            Some(instants[0]),
        );

        vec![peer0, peer1, peer2, peer3, peer4, peer5, peer6, peer7, peer8, peer9]
    }

    #[test]
    fn test_eviction_ranks() {
        let ranks = EvictionRanks {
            ip_prefix_bucket: 4.0,
            time_connected: 1.0,
            last_ping_duration: 0.0,
            last_block_transfer: 2.0,
            last_tx_transfer: 3.0,
        };
        assert_eq!(ranks.lowest_rank(), 0.0);
        assert_eq!(ranks.highest_none_latency_rank(), 4.0);

        let ranks = EvictionRanks {
            ip_prefix_bucket: 3.0,
            time_connected: 0.0,
            last_ping_duration: 4.0,
            last_block_transfer: 1.0,
            last_tx_transfer: 2.0,
        };
        assert_eq!(ranks.lowest_rank(), 0.0);
        assert_eq!(ranks.highest_none_latency_rank(), 3.0);

        let ranks = EvictionRanks {
            ip_prefix_bucket: 2.0,
            time_connected: 4.0,
            last_ping_duration: 3.0,
            last_block_transfer: 0.0,
            last_tx_transfer: 1.0,
        };
        assert_eq!(ranks.lowest_rank(), 0.0);
        assert_eq!(ranks.highest_none_latency_rank(), 4.0);

        let ranks = EvictionRanks {
            ip_prefix_bucket: 1.0,
            time_connected: 3.0,
            last_ping_duration: 2.0,
            last_block_transfer: 4.0,
            last_tx_transfer: 0.0,
        };
        assert_eq!(ranks.lowest_rank(), 0.0);
        assert_eq!(ranks.highest_none_latency_rank(), 3.0);

        let ranks = EvictionRanks {
            ip_prefix_bucket: 0.0,
            time_connected: 2.0,
            last_ping_duration: 1.0,
            last_block_transfer: 3.0,
            last_tx_transfer: 4.0,
        };
        assert_eq!(ranks.lowest_rank(), 0.0);
        assert_eq!(ranks.highest_none_latency_rank(), 2.0);
    }

    #[test]
    fn test_eviction_iter_from_peers() {
        let test_peers = build_test_peers();
        let test_peers = test_peers.iter().collect::<Vec<_>>();
        let eviction_iter_vec = from_peers(&test_peers).collect::<Vec<(&Peer, EvictionRanks)>>();

        let expected_ranks = vec![
            EvictionRanks {
                ip_prefix_bucket: 0.0,
                time_connected: 1.0,
                last_ping_duration: 2.0,
                last_block_transfer: 3.0,
                last_tx_transfer: 4.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 1.5,
                time_connected: 0.0,
                last_ping_duration: 1.0,
                last_block_transfer: 2.0,
                last_tx_transfer: 3.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 1.5,
                time_connected: 2.0,
                last_ping_duration: 0.0,
                last_block_transfer: 1.0,
                last_tx_transfer: 2.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 4.0,
                time_connected: 3.0,
                last_ping_duration: 3.0,
                last_block_transfer: 0.0,
                last_tx_transfer: 1.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 4.0,
                time_connected: 4.0,
                last_ping_duration: 4.0,
                last_block_transfer: 4.0,
                last_tx_transfer: 0.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 4.0,
                time_connected: 5.0,
                last_ping_duration: 5.0,
                last_block_transfer: 5.0,
                last_tx_transfer: 5.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 7.5,
                time_connected: 6.0,
                last_ping_duration: 6.0,
                last_block_transfer: 6.0,
                last_tx_transfer: 8.5,
            },
            EvictionRanks {
                ip_prefix_bucket: 7.5,
                time_connected: 7.0,
                last_ping_duration: 7.0,
                last_block_transfer: 7.0,
                last_tx_transfer: 8.5,
            },
            EvictionRanks {
                ip_prefix_bucket: 7.5,
                time_connected: 8.0,
                last_ping_duration: 8.0,
                last_block_transfer: 8.5,
                last_tx_transfer: 6.0,
            },
            EvictionRanks {
                ip_prefix_bucket: 7.5,
                time_connected: 9.0,
                last_ping_duration: 9.0,
                last_block_transfer: 8.5,
                last_tx_transfer: 7.0,
            },
        ];

        assert_eq!(eviction_iter_vec.len(), expected_ranks.len());
        for (i, (_, ranks)) in eviction_iter_vec.iter().enumerate() {
            assert_eq!(ranks, &expected_ranks[i]);
        }
    }

    #[test]
    fn test_eviction_iter_filter_peers() {
        let test_peers = build_test_peers();
        let test_peers = test_peers.iter().collect::<Vec<_>>();
        let iterations = test_peers.len();
        let eviction_iter_vec = from_peers(&test_peers).collect::<Vec<(&Peer, EvictionRanks)>>();

        for i in 0..iterations + 1 {
            let mut removed_counter = HashMap::<u64, usize>::new();
            let mut filtered_counter = HashMap::<u64, usize>::new();
            let eviction_candidates_iter = eviction_iter_vec.clone().into_iter();
            let filtered_eviction_set = eviction_candidates_iter.filter_peers(i, cmp_strats::by_lowest_rank).collect_vec();
            let removed_eviction_set =
                eviction_iter_vec.clone().into_iter().filter(|item| !filtered_eviction_set.contains(item)).collect_vec();
            assert_eq!(filtered_eviction_set.len(), iterations - i);
            assert_eq!(removed_eviction_set.len(), i);
            for (_, er) in &removed_eviction_set {
                *removed_counter.entry(er.lowest_rank().to_bits()).or_insert(0) += 1;
            }
            for (_, er) in &filtered_eviction_set {
                *filtered_counter.entry(er.lowest_rank().to_bits()).or_insert(0) += 1;
            }
            match i {
                0 => {
                    assert_eq!(removed_counter.len(), 0);

                    assert_eq!(filtered_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 4);
                }
                1 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter.len(), 1);

                    assert_eq!(filtered_counter[&0.0_f64.to_bits()], 4);
                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 4);
                }
                2 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 2);
                    assert_eq!(removed_counter.len(), 1);

                    assert_eq!(filtered_counter[&0.0_f64.to_bits()], 3);
                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 4);
                }
                3 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 3);
                    assert_eq!(removed_counter.len(), 1);

                    assert_eq!(filtered_counter[&0.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 4);
                }
                4 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 4);
                    assert_eq!(removed_counter.len(), 1);

                    assert_eq!(filtered_counter[&0.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 4);
                }
                5 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter.len(), 1);

                    assert_eq!(filtered_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 3);
                }
                6 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter.len(), 2);

                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 2);
                }
                7 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter.len(), 3);

                    assert_eq!(filtered_counter[&6.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 2);
                }
                8 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(removed_counter.len(), 3);

                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(filtered_counter.len(), 1);
                }
                9 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(removed_counter[&7.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter.len(), 4);

                    assert_eq!(filtered_counter[&7.0_f64.to_bits()], 1);
                    assert_eq!(filtered_counter.len(), 1);
                }
                10 => {
                    assert_eq!(removed_counter[&0.0_f64.to_bits()], 5);
                    assert_eq!(removed_counter[&4.0_f64.to_bits()], 1);
                    assert_eq!(removed_counter[&6.0_f64.to_bits()], 2);
                    assert_eq!(removed_counter[&7.0_f64.to_bits()], 2);
                    assert_eq!(removed_counter.len(), 4);

                    assert_eq!(filtered_counter.len(), 0);
                }
                _ => panic!("unexpected i value"),
            }
        }
    }

    #[test]
    fn test_eviction_iter_select_peers_weighted() {
        try_init_logger("info");
        let test_peers = build_test_peers();
        let test_peers = test_peers.iter().collect::<Vec<_>>();
        let eviction_iter_vec = from_peers(&test_peers).collect::<Vec<(&Peer, EvictionRanks)>>();
        let total_weight = 59.5;
        let expected_probabilities = vec![
            // we add one to avoid 0, and nan numbers.. `weight_strats::by_highest_none_latency_rank` adds 1 to the rank for these hypothetical situation.
            (1.0 + 1.0) / total_weight,
            (1.5 + 1.0) / total_weight,
            (2.0 + 1.0) / total_weight,
            (4.0 + 1.0) / total_weight * 2.0, // we have two 4.0 ranks
            (5.0 + 1.0) / total_weight,
            (7.5 + 1.0) / total_weight * 2.0, // we have two 7.5 ranks
            (8.0 + 1.0) / total_weight,
            (9.0 + 1.0) / total_weight,
        ];
        assert_eq!(expected_probabilities.iter().sum::<f64>(), 1.0);
        assert_eq!(expected_probabilities.len(), 8);

        let mut selected_counter = HashMap::<u64, f64>::new();
        let num_of_trials = 2054;
        for _ in 0..num_of_trials {
            //println!("sample_size: {}", sample_size);
            let eviction_iter = eviction_iter_vec.clone().into_iter();
```